src/cljs/vanilla/widgets/carousel.cljs

```diff
@@ -147,96 +149,206 @@
+            {:display-name (str "Spectrum slide " idx)
 
              :reagent-render
-             (fn [args]
-               [:div#chartSlide {:style {:width "100%" :height "100%"}}])
+                           (fn [args]
+                             [:div#chartSlide {:style {:width "100%" :height "100%"}}])
 
              :component-did-mount
-             (fn [this]
-               ^{:key idx} [:> (Highcharts/chart (r/dom-node this) (clj->js s))])}))))))
+                           (fn [this]
+                             ^{:key idx} [:> (Highcharts/chart (r/dom-node this) (clj->js s))])}))))))
 
 
 
 (defn make-sankey-widget [name data options]
   ; TODO: this is a hack for the following hack (does NOT unsubscribe to sources when widget closes)
   (ds/data-source-subscribe [:signal-path-service])
 
-  (let [data @(rf/subscribe [:app-db :signal-path-service])
+  (let [data         @(rf/subscribe [:app-db :signal-path-service])
         chart-config @(rf/subscribe [:hc-type :sankey-chart])
-        sources (get-in data [:data :series])]
+        sources      (get-in data [:data :series])]
+
+    ;(prn "Making sankey carousel widget: " data " //// " sources)
+    ;  " //// chart-config " chart-config)
+
+    (carousel
+      (for [[idx s] (map-indexed vector sources)]
+        (do
+          ;(prn "make-sankey-chart " s)
+          (r/create-element
+            ^{:key idx} (m/make-chart chart-config
+                          {:data {:series [s]}}
+                          {:series {:showInLegend true :visible false}})))))))
+
+
+(defn make-heatmap-widget [name data options]
+  ; TODO: this is a hack for the following hack (does NOT unsubscribe to sources when widget closes)
+  (ds/data-source-subscribe [:heatmap-data])
+
+  (let [data         @(rf/subscribe [:app-db :heatmap-data])
+        chart-config @(rf/subscribe [:hc-type :heatmap-chart])
+        sources      (get-in data [:data :series])]
 
     ;(prn "Making sankey carousel widget: " data " //// " sources)
     ;  " //// chart-config " chart-config)
 
     (carousel
       (for [[idx s] (map-indexed vector sources)]
-       (do
-         ;(prn "make-sankey-chart " s)
-         (r/create-element
-           ^{:key idx} (m/make-chart chart-config
-                         {:data {:series [s]}}
-                         {:series {:showInLegend true :visible false}})))))))
+        (do
+          ;(prn "make-sankey-chart " s)
+          (r/create-element
+            ^{:key idx} (m/make-chart chart-config
+                          {:data {:series [s]}}
+                          {:series {:showInLegend true :visible false}})))))))
+
+
+(defn- build-charts [name labels axis-types data]
+  (map (fn [{d :data :as all}]
+         (let [cvtAll (assoc all :data d)]
+           (do
+             (prn "build-charts " cvtAll)
+             (big-chart name labels axis-types cvtAll))))
+    data))
+
+
+
+(defn make-telemetry-widget [name data options]
```

Comment:
Do we want to genericize these carousel widget functions?  I know we need to re-architect the overall widget/carousel creation structure, but in the meantime, this is a lot of duplicate code.   It does seem like theres just enough differences in sources and titles and such between them to justify, just thinking.

Reply:
I'm pretty sure widget-workshop will sort this all out since it's a design goal to do just that, among many other things about widgets and data-sources, but this repetition shows that such a re-think is really, really necessary.